lib/Models/UserDrawing.js

```diff
@@ -0,0 +1,418 @@
+var defined = require('terriajs-cesium/Source/Core/defined');
+var defaultValue = require('terriajs-cesium/Source/Core/defaultValue');
+var Color = require('terriajs-cesium/Source/Core/Color');
+var PolylineGlowMaterialProperty = require('terriajs-cesium/Source/DataSources/PolylineGlowMaterialProperty');
+var CustomDataSource = require('terriajs-cesium/Source/DataSources/CustomDataSource');
+var CallbackProperty = require('terriajs-cesium/Source/DataSources/CallbackProperty');
+var PolygonHierarchy = require('terriajs-cesium/Source/Core/PolygonHierarchy');
+var knockout = require('terriajs-cesium/Source/ThirdParty/knockout');
+var Entity = require('terriajs-cesium/Source/DataSources/Entity.js');
+var when = require('terriajs-cesium/Source/ThirdParty/when');
+var Cartesian3 = require('terriajs-cesium/Source/Core/Cartesian3');
+
+/**
+ * Callback for when a point is clicked.
+ * @callback onPointClicked
+ * @param {CustomDataSource} customDataSource Contains all point entities that user has selected so far
+ */
+
+/**
+ * Callback for when clean up is happening, i.e., for done or cancel.
+ * @callback onCleanUp
+ */
+
+/**
+ * Callback for when the dialog is displayed, to provide a custom message
+ * @callback onMakeDialogMessage
+ * @return {String} Message to add to dialog
+ */
+
+/**
+ * @alias UserDrawing
+ * @constructor
+ *
+ * For user drawings, which includes lines and/or a polygon
+ * @param {Object} options Object with the following properties, all of which are optional except terria:
+ * @param {Terria} [options.terria] The Terria instance.
+ * @param {String} [options.messageHeader] Heading for the dialog which pops up when in user drawing mode
+ * @param {Bool}   [options.allowPolygon] Let the user click on first point to close loop
+ * @callback {onPointClicked} options.onPointClickedCallback Optional way to subscribe to point clicks
+ * @callback {onCleanUp} options.onCleanUpCallback Optional way to add own cleanup
+ * @callback {onMakeDialogMessage} options.onMakeDialogMessageCallback Optional way to customise dialog message
+ */
+var UserDrawing = function(options) {
+
+    options = defaultValue(options, defaultValue.EMPTY_OBJECT);
+    if (!defined(options.terria)) {
+        throw new DeveloperError('Terria instance is required.');
+    }
+
+    /**
+     * Text that appears at the top of the dialog when drawmode is active.
+     * @type {String}
+     * @default "Draw on Map"
+     */
+    this.messageHeader = defaultValue(options.messageHeader, "Draw on Map");
+
+    /**
+     * If true, user can click on first point to close the line, turning it into a polygon.
+     * @type {Bool}
+     * @default true
+     */
+    this.allowPolygon = defaultValue(options.allowPolygon, true);
+
+    /**
+     * Callback that occurs when point is clicked (may be added or removed). Function takes a CustomDataSource which is
+     * a list of PointEntities.
+     * @type {function}
```

Comment:
`@type` should be `{onPointClicked}`.  Actually my personal preference is to avoid `on` in the type of a callback function.  Something like `PointClickedCallback` is a good name IMO.  `onPointClicked` is fine for the name of the property itself.